syntax = "proto3";

package authservice;

import "google/protobuf/timestamp.proto";

service AuthService {
    // Authentication Operations
    rpc Register (RegisterRequest) returns (AuthResponse);
    rpc Login (LoginRequest) returns (AuthResponse);
    rpc RefreshToken (RefreshTokenRequest) returns (AuthResponse);
    rpc Logout (LogoutRequest) returns (LogoutResponse);
    
    // User Management
    rpc GetUserProfile (GetUserRequest) returns (UserProfileResponse);
    rpc UpdateUserProfile (UpdateUserRequest) returns (UserProfileResponse);
    rpc DeleteUser (DeleteUserRequest) returns (DeleteUserResponse);
    
    // Token Operations
    rpc ValidateToken (ValidateTokenRequest) returns (ValidateTokenResponse);
    rpc RevokeToken (RevokeTokenRequest) returns (RevokeTokenResponse);

    // Health Check
    rpc CheckHealth(HealthCheckRequest) returns (HealthCheckResponse) {}
}

// Authentication Messages
message RegisterRequest {
    string email = 1;
    string password = 2;
    string display_name = 3;
    string role = 4;
}

message LoginRequest {
    string email = 1;
    string password = 2;
}

message RefreshTokenRequest {
    string refresh_token = 1;
}

message LogoutRequest {
    string access_token = 1;
}

message LogoutResponse {
    bool success = 1;
    string message = 2;
}

// User Profile Messages
message GetUserRequest {
    string user_id = 1;
}

message UpdateUserRequest {
    string user_id = 1;
    string display_name = 2;
    string email = 3;
    string role = 4;
    map<string, string> custom_claims = 5;
}

message UserProfileResponse {
    string user_id = 1;
    string email = 2;
    string display_name = 3;
    string role = 4;
    map<string, string> custom_claims = 5;
    google.protobuf.Timestamp created_at = 6;
    google.protobuf.Timestamp updated_at = 7;
}

// Token Messages
message ValidateTokenRequest {
    string token = 1;
}

message ValidateTokenResponse {
    bool is_valid = 1;
    string user_id = 2;
    map<string, string> claims = 3;
    string error_message = 4;
}

message RevokeTokenRequest {
    string token = 1;
}

message RevokeTokenResponse {
    bool success = 1;
    string message = 2;
}

// Common Response
message AuthResponse {
    string access_token = 1;
    string refresh_token = 2;
    string user_id = 3;
    string email = 4;
    string display_name = 5;
    string role = 6;
    map<string, string> custom_claims = 7;
    string error_message = 8;
}

// Delete User
message DeleteUserRequest {
    string user_id = 1;
}

message DeleteUserResponse {
    bool success = 1;
    string message = 2;
}

// Health Check Messages
message HealthCheckRequest {}

message HealthCheckResponse {
    string status = 1;  // SERVING, NOT_SERVING, UNKNOWN
    string message = 2;
} 